admin/worker/run_autoscaler.go

```diff
@@ -18,10 +20,29 @@ func (w *Worker) runAutoscaler(ctx context.Context) error {
 	}
 
 	for _, rec := range recs {
-		// TODO: Add autoscaling logic based on the recommendation here.
-		// Consider checking rec.UpdatedOn to avoid making autoscaling decision on stale data!
-		w.logger.Info("autoscaler recommendation", zap.String("project_id", rec.ProjectID), zap.Int("recommended_slots", rec.RecommendedSlots))
-		break
+		duration := time.Since(rec.UpdatedOn)
+		if duration < 24*time.Hour {
+			w.logger.Debug("skipping autoscaler: the project has been scaled recently", zap.String("project_id", rec.ProjectID), zap.Time("project_updated_on", rec.UpdatedOn))
+			break
+		}
+
+		targetProject, err := w.admin.DB.FindProject(ctx, rec.ProjectID)
+		if err != nil {
+			w.logger.Warn("failed to find project:", zap.String("project_id", rec.ProjectID), zap.Error(err))
+			break
+		}
+
+		opts := &database.UpdateProjectOptions{
+			ProdSlots: rec.RecommendedSlots,
+		}
```

Comment:
I think this is a great concern 👍🏼  and have also talked with @kaspersjo :

As the essence to have a careful control here is knowing _**when to scale**_ and _**how to scale**_:
- On one hand, we want to avoid frequent or unnecessary scaling, such as slots from `30 -> 31` or `60 -> 58`.
- On the other hand, we need the scaler to be responsive when "necessary"

@kaspersjo raised the idea to have a fixed number of slots difference (`abs(current_slots - recommend_slots) <= min_slots_scale_limit`), which can be clearer than a percentage threshold.

Well, from a double thought, I'd prefer a percentage threshold (`min_slots_scale_percentage` than a fix number of scale limit `min_slots_scale_limit`:
- Imagine we have two projects with different scales (say `A` with 100 slots and `B` with 5 slots). As we want to reduce unnecessary scaling that leads to downtime, defining a literal `min_slots_scale_limit` is hard:
    - If it is defined small, like `1-20`, it does not help control `A`.
    - And it is hard to find a "bigger" number, say `20`, which maybe fit for `A`, but probably won't make sense for a project which can be scaled from `20` to `40`.
- So there is a balance to maintain, and I propose to define `min_slots_scale_percentage`, like `10%`:
    - It should be useful for bigger projects (slots >= 50)
    - For small projects, it is a tradeoff, but I think it is acceptable, assuming restarting a small project is easier/quicker than restarting a bigger project.

Reply:
I like your approach here. Having a percentage also allows us to scale at the small end, e.g. if a project is very small, we can scale down from 2 slots (default) to 1 slot.